### ISSUE TITLE
broken `pip check` for tensorboard affects all downstream packages

### ISSUE THREAD

**Issue by h-vetinari** [conda-forge/tensorboard-feedstock#65]
Not only does this PR break `pip check` for the tensorboard recipe, it also [breaks `pip check`](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=565471&view=logs&j=656edd35-690f-5c53-9ba3-09c10d0bea97&t=e5c8ab1d-8ff9-5cae-b332-e15ae582ed2d) for every package that depends on tensorboard, requiring them to either remove `pip check` or [pin protobuf](https://github.com/conda-forge/torchgeo-feedstock/pull/8/commits/53275cbf13f427354a5813d3be52d4895f66ac88). Is there any reason for this recipe to have a different set of dependency version requirements than the published wheel?

_Originally posted by @adamjstewart in https://github.com/conda-forge/tensorboard-feedstock/issues/64#issuecomment-1241356362_

**Comment by h-vetinari** [comment]
Yes there is a reason. Upstream tensorboard/tensorflow cannot take into account non-python artefacts like shared libraries, and so they have to be bundled into each of the wheela, but also there needs to be a great amount of care to ensure that no incompatible (shared) libraries get installed from different packages.

Conda-forge _can_ take care of non-python dependencies, and so we can have just _one_ shared library (e.g. libprotobuf) that gets used by many packages. To do that, we need to ensure that all packages depending on libprotobuf are built against a consistent version across the whole ecosystem (otherwise packages become impossible to co-install, because you cannot have two different versions of the same shared library in an environment).

This happens constantly not only for protobuf but for a very large number of libraries that underlie a lot of the ecosystem, check out conda-forge.org/status under "other migrations".


Furthermore, there is no fundamental incompatibility of the tensorboard source code with protobuf>3.19, just that we're faster than upstream in getting the package rebuilt for that. Not only are we faster, but we also (in general) need to keep moving forward, because it becomes impossible to reasonably maintain more than one packages per library (except perhaps for the duration of a migration), because it would mean a combinatorial explosion of packages needing to be built.

So long story short, we need to move to 3.21 to keep up with the rest of the conda-forge ecosystem (and if you pin to 3.19, the packages of that feedstock will likewise become very hard to co-install with other packages).

Normally this is easy to fix with a patch to the respective package, but here we're in the unusual and less-than-ideal situation of repackaging a wheel.

We could try to patch the wheel metadata itself (which would be the fastest solution, assuming there's a good way to do that), or start building from source.

**Comment by h-vetinari** [comment]
@conda-forge/core
Does anyone have experience with patching wheel metadata? It's very much not ideal that this feedstock is repackaging a wheel, but tensorflow packages often have circular dependencies with the main package, so not sure there's a fast way to start building from source.

**Comment by hmaarrfk** [comment]
I have removed the dependencies altogether from the setup.py file before

https://github.com/conda-forge/zaber-motion-feedstock/blob/main/recipe/filter_setup.py

**Comment by hmaarrfk** [comment]
For similar reasons

**Comment by h-vetinari** [comment]
Thanks @hmaarrfk, that looks pretty good! In that case you still had the sources though, so here we'd first have to unpack the wheel (shouldn't be too hard, it's just an archive+metadata), modify the setup.py & then install.

@adamjstewart @ngam, anyone willing to give this a shot?

**Comment by hmaarrfk** [comment]
I would try to modify this recipe to not use the wheel.

Pip check has always been a luxury. I have a few comments where I explicitly call out the tensorflow because of this issue.

**Comment by h-vetinari** [comment]
> I would try to modify this recipe to not use the wheel.

Sure, but that might very well end up being a very long undertaking (especially with possible bootstrap cycles; and I don't have time to dedicate to this ATM), so it very likely doesn't qualify as a quick fix.

**Comment by adamjstewart** [comment]
I would recommend not trying to build from source. I'm a maintainer for the tensorboard recipe in Spack. We used to build from source, but it requires Bazel, and the build takes upwards of 30 min. The build also doesn't work on systems like Apple M1. We eventually switched to using a wheel. But if you want to go down that route, I can share the appropriate build command.

Has anyone tried to update the upstream repo to mark the fact that newer protobuf works?

**Comment by h-vetinari** [comment]
We're fine in principle with bazel + cross compilation, we even do that for tensorflow itself. But if there's no circular dependence with tf, then it should be fine to switch to source.

> Has anyone tried to update the upstream repo to mark the fact that newer protobuf works?

It's theoretically possible, but upstream has a totally different distribution model. They'll get there eventually, but I suspect they won't change the protobuf range to suit anyone's but their own needs.

**Comment by adamjstewart** [comment]
FWIW, this was our build recipe for building tensorboard from source:
https://github.com/spack/spack/blob/1bd33d88bd5bdf6820800b71b2f5f798e857920b/var/spack/repos/builtin/packages/py-tensorboard/package.py
It required a lot of patching to build, just like TF.

**Comment by hmaarrfk** [comment]
You could just open the file:
```
sed -i '/Requires-Dist/d' ${SP_DIR}/tensorboard-{{ version }}.dist-info/METADATA
```
in the recipe, and explain why.

The original metadata file contains:
```
Metadata-Version: 2.1
Name: tensorboard
Version: 2.10.0
Summary: TensorBoard lets you watch Tensors Flow
Home-page: https://github.com/tensorflow/tensorboard
Author: Google Inc.
Author-email: packages@tensorflow.org
License: Apache 2.0
Keywords: tensorflow tensorboard tensor machine learning visualizer
Classifier: Development Status :: 4 - Beta
Classifier: Intended Audience :: Developers
Classifier: Intended Audience :: Education
Classifier: Intended Audience :: Science/Research
Classifier: License :: OSI Approved :: Apache Software License
Classifier: Programming Language :: Python :: 3
Classifier: Programming Language :: Python :: 3.6
Classifier: Programming Language :: Python :: 3.7
Classifier: Programming Language :: Python :: 3.8
Classifier: Programming Language :: Python :: 3.9
Classifier: Programming Language :: Python :: 3 :: Only
Classifier: Topic :: Scientific/Engineering :: Mathematics
Classifier: Topic :: Software Development :: Libraries :: Python Modules
Classifier: Topic :: Software Development :: Libraries
Requires-Python: >=3.6
License-File: LICENSE
Requires-Dist: absl-py (>=0.4)
Requires-Dist: grpcio (>=1.24.3)
Requires-Dist: google-auth (<3,>=1.6.3)
Requires-Dist: google-auth-oauthlib (<0.5,>=0.4.1)
Requires-Dist: markdown (>=2.6.8)
Requires-Dist: numpy (>=1.12.0)
Requires-Dist: protobuf (<3.20,>=3.9.2)
Requires-Dist: requests (<3,>=2.21.0)
Requires-Dist: setuptools (>=41.0.0)
Requires-Dist: tensorboard-data-server (<0.7.0,>=0.6.0)
Requires-Dist: tensorboard-plugin-wit (>=1.6.0)
Requires-Dist: werkzeug (>=1.0.1)
Requires-Dist: wheel (>=0.26)

TensorBoard is a suite of web applications for inspecting and understanding
your TensorFlow runs and graphs.

Releases prior to 1.6.0 were published under the ``tensorflow-tensorboard`` name
and may be found at https://pypi.python.org/pypi/tensorflow-tensorboard.
```

and remove the Requires-Dist lines. Like all of them.

**Comment by hmaarrfk** [comment]
explain the hack

**Comment by hmaarrfk** [comment]
the torchgeo package looks ok. The pinnings seem conservative, but many that manage large applications find this to be a good approach since it gives them the ability to test new packages before they come out. Up to them....

PRs welcome that would help `pip check` here, even if from outside them maintainers.

**Comment by adamjstewart** [comment]
> Please consider relaxing your very strict pins if you want people to productively use your package.

Our pins are already fairly relaxed. The minimum is tested in CI, and can't go any lower. The maximum is based on [semver](https://semver.org/). Since we aren't able to test new major releases, which are used only when backwards incompatible changes are made, we set an upper bound on the newest major version that has been tested.

Unfortunately, package managers like pip and conda do not allow you to change the version bounds of historical releases. [Spack](https://spack.io) does, but not everyone uses Spack, so we try to be conservative about releases we've never tested before. This was needed because new major releases of libraries like pillow, and new minor releases of beta software like torchmetrics and radiant-mlhub, have resulted in incompatibilities on multiple occasions. If you recall, when protobuf 4 was released, pretty much all libraries that depend on protobuf broke since they didn't pin their dependencies to be `<4` like they should have. It took months for these libraries to ship new releases to fix this mistake. I would rather avoid a repeat of https://github.com/protocolbuffers/protobuf/issues/10051.

> The geo ecosystem is really fragile and annoying, and your package makes it even more so.

I would argue that relaxing our pins even further (aka removing them) would make the geo ecosystem significantly more fragile, since users could easily install versions of dependencies that simply do not work. I've never encountered any issues installing torchgeo, although I don't personally use conda.

**Comment by adamjstewart** [comment]
I wasn't aware of pangeo-forge, neat! If you encounter any future issues with torchgeo just let me know. We try to keep up-to-date with all new major releases, although we don't yet support the latest radiant-mlhub release. And major releases don't happen very often, so we rarely need to bump upper bounds.

But we aren't pinning protobuf, tensorboard is. I'm happy to submit a PR to bump the max dependency version of tensorboard if no one else wants to.

**Comment by isuruf** [comment]
> Unfortunately, package managers like pip and conda do not allow you to change the version bounds of historical releases.

conda does. It's not easy or obvious though. See for eg: https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/blob/main/recipe/gen_patch_json.py#L541-L548

**Comment by adamjstewart** [comment]
> You always end up bringing spack

I'm [biased](https://github.com/spack/spack/graphs/contributors) 😄 

> You should check out conda btw, you may end up in love

I used to use conda for ~5 years but eventually gave up. Years of fighting against ABI compatibility issues every time I mixed channels or upgraded packages was a strong deterrent. Conda also lacks many of the features I need, including:

* optional dependencies
* optional build flags
* control over compilers
* control over MPI and BLAS/LAPACK libraries

With Spack I can do something like:
```console
$ spack install py-torchgeo+datasets %gcc@9.3.0 ^py-torch@1.11.0+cuda+distributed+mpi cuda_arch=80 ^openmpi
```
This builds TorchGeo, with optional dataset-specific dependencies, using PyTorch 1.11, with CUDA and MPI support, using the OpenMPI backend, all with the GCC 9.3.0 compiler. 

Of course, this is only possible because Spack builds everything from source, and there are many disadvantages of from-source package managers (especially build time). I don't want to turn this into a discussion of which package manager is better (especially in an away game) since they serve very different purposes. Conda is more targeted to the average user, while Spack is more targeted to supercomputing environments where microarchitecture-specific optimizations are important for optimal performance.

Thanks for #67, I'll try removing our protobuf pin and see if pip check now passes. I'll also do what I can to encourage TF/tensorboard devs to offer support for protobuf 3.20. I think 4.21 support would require work on their part, but 3.20 support should work fine.
